lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,444 @@
+    PREFIX_STATUS_INDICATORS = "prefix_status_indicators"
+    MIALIST = "mialist"
+    EMPTY_CONTROLLERS = "empty_controllers"
+    UNEXPECTED_CONTROLLERS = "unexpected_controllers"
+    INVALID_TB_DIRS = "invalid_tb_dirs"
+    EMPTY_TARBALL_DIRS = "empty_tarball_dirs"
+    INVALID_UNPACKING_DIRS = "invalid_unpacking_dirs"
+    TARBALL_LINKS = "tarball_links"
+    INVALID_TB_LINKS = "invalid_tb_links"
+    INCORRECT_TB_DIR_LINKS = "incorrect_tb_dir_links"
+    INVALID_TB_DIR_LINKS = "invalid_tb_dir_links"
+    UNUSED_PREFIX_FILES = "unused_prefix_files"
+    MISSING_PREFIX_FILES = "missing_prefix_files"
+    BAD_PREFIX_FILES = "bad_prefix_files"
+    BAD_PREFIXES = "bad_prefixes"
+    UNEXPECTED_USER_LINKS = "unexpected_user_links"
+    WRONG_USER_LINKS = "wrong_user_links"
+    INCONCEIVABLE = "inconceivable"
+    ERROR = "error"
+
+    def __init__(self, name, path, config):
+        self.name = name
+        self.path = path
+        self.config = config
+        self.controllers = list()
+        self.bad_controllers = list()
+        self.validation_list = None
+
+    def add_controller(self, controller):
+        self.controllers.append(controller)
+
+    def add_bad_controller(self, controller):
+        self.bad_controllers.append(controller)
+
+    def add_error_or_inconceivable_entries(self, val, controller, dirs):
+        self.validation_list[val][controller] = dirs
+
+    def add_unexpected_entries(self, val, controller, dirs):
+        """Gather values for each validation_list dict key"""
+        self.validation_list[val][controller] = dirs
+
+    def header(self, fp, status):
+        """Adds main starting and ending messages for each Hierarchy"""
+        lead_newline = "\n" if status == "start" else ""
+        fp.write(
+            f"{lead_newline}\n{status}-{self.config.TS[4:]}: {self.name} hierarchy: {self.path}\n"
+        )
+        return
+
+    def check_controller_in_list(self, controller):
+        """Checks whether there are controllers in dictionary or not"""
+        return any(
+            [
+                controller in self.validation_list[key]._dict
+                for key in self.validation_list
+            ]
+        )
+
+    """
+    ##FIXME: The mention of hierarchies(archive, controller) from here on, in
+            Hierarchy class is mainly to keep the format algned with the shell script
+    """
+
+    def dump_check(self, fp, controller, key, cnt, hierarchy=False):
+        """Output Messages for each list of controllers, files,
+        directories, etc depicting their cause of being added to
+        that list or dictionary
+        """
+        lead_asterisk = "\t* " if hierarchy == "archive" else "\t"
+        if controller in self.validation_list[key]._dict:
+            if key not in ["subdir_status_indicators", "prefix_status_indicators"]:
+                fp.write(f"{lead_asterisk}{self.validation_list[key]._get_msg()}\n")
+                if hierarchy == "archive":
```

Comment:
The default value for `hierarchy` is `False`, suggesting a boolean. This is a weird convention. Presumably this parameter is based on the subclass identity, and should be encapsulated in the class hierarchy rather than being an explicit argument. Should the default at least be "" or None, with a string type? Or is the intended protocol more complex? (In which case it should at least be documented in the docstring.)

Reply:
ok sure